add the ability to load images using ajax

robertleeplummerjr opened the pull request:
Added the ability to use ajax, which can overcome certain security deficiencies found in Chrome and IE.

Overcomes the following obstacle: http://stackoverflow.com/questions/19244006/chrome-basic-auth-image-requests-return-401-but-not-when-called-directly

This is now a browser behaviour, where other browsers are following suite.

robertleeplummerjr committed:
Added the ability to use ajax, which can overcome certain security deficiencies found in Chrome and IE

tmcw commented on index.js:
```diff
@@ -208,6 +208,13 @@ module.exports = function leafletImage(map, callback) {
+            });
+            return;
+        }
         im.src = url;
```
This should be an else, right? Otherwise this both loads images via URL and loads them with ajax slightly after.

robertleeplummerjr replied:
It is an implied else, with the `return;` at the bottom of the if statement.

robertleeplummerjr replied:
I corrected the else to me more explicit.  Also I fixed a bug that caused onload to be called multiple times depending on latency and I also now cache the images internally to cut down on requests.

robertleeplummerjr committed:
create image cache when using ajax
use else because it looks better
correct onload duplicate callback issue causing images to intermittently show up

robertleeplummerjr committed:
space correction

robertleeplummerjr committed:
space correction

robertleeplummerjr commented:
Is this pr satisfactory?  I'd like to get it merged and start on handling canvas layers so that, for example, the points here: http://robertleeplummerjr.github.io/Leaflet.glify/ show up.  I plan on adding this today.

robertleeplummerjr committed:
add the ability to render a canvas directly as image if render method exists and canvas property exists

robertleeplummerjr committed:
spacing

robertleeplummerjr commented:
added canvas layer rendering

tmcw commented on index.js:
```diff
@@ -202,18 +212,74 @@ module.exports = function leafletImage(map, callback) {
 
-        if (isBase64) im.onload();
+            if (isBase64 && !loaded) {
+                setTimeout(function () {
```
why is there a settimeout here, and previously?

robertleeplummerjr replied:
It wouldn't work without it.  Honestly not sure.

robertleeplummerjr replied:
There is always that chance that it had something to do with the gravitation and alignment of the planets...  I'll try it later without it just to make sure.

:P

tmcw commented:
Happy to merge once the setTimeout is deciphered.

robertleeplummerjr commented on leaflet-image.js:
```diff
@@ -204,20 +213,76 @@ module.exports = function leafletImage(map, callback) {
+                im.src = cache;
+                if (!loaded) {
+                    setTimeout(function() {
+                        im.onload();
```
The image isn't yet added to the page, but is base64.  `setTimeout(fn, 0)` ensures it is loaded when the next available context is, which (since we just set the src) is just after the src is rendered.  Otherwise the image is not yet rendered, thus resulting in a blank image.

robertleeplummerjr replied:
Sorry for the tardy response.

robertleeplummerjr committed:
Merge remote-tracking branch 'upstream/gh-pages' into gh-pages

# Conflicts:
#	index.js
#	leaflet-image.js

robertleeplummerjr committed:
Merge remote-tracking branch 'upstream/gh-pages' into gh-pages

# Conflicts:
#	index.js
#	leaflet-image.js

robertleeplummerjr committed:
cleanup, post merge

robertleeplummerjr committed:
adding Uint8Array to globals

robertleeplummerjr commented:
To restate: https://github.com/mapbox/leaflet-image/pull/51/files/655589ad72b3da6e8ad5e71ad3dfc5c8467a466c#r66086754
